src/keymap.cpp

```diff
@@ -783,122 +1024,50 @@ void keyProcessMappings(bool bExclude)
+	/* If mappings have been updated, sort the mappings by whether or not they have meta keys */
+	if (bMappingsSortOrderDirty)
+	{
+		keyMappings.sort([](const KEY_MAPPING& a, const KEY_MAPPING& b) {
+			// Sort by meta. This causes all mappings with meta to be checked before non-meta mappings,
+			// avoiding having to check for meta-conflicts in the processing loop. (e.g. if we should execute
+			// a mapping with right arrow key, depending on if another binding on shift+right-arrow is executed
+			// or not). In other words, if any mapping with meta is executed, it will consume the respective input,
+			// preventing any non-meta mappings with the same input from being executed.
+			return a.hasMeta() && !b.hasMeta();
+		});
+		bMappingsSortOrderDirty = false;
+	}
+	std::vector<KeyMappingInput> consumedInputs;
 
-	/* Run through all our mappings */
-	for (auto keyToProcess = keyMappings.begin(); keyToProcess != keyMappings.end(); ++keyToProcess)
+	/* Run through all sorted mappings */
+	for (const KEY_MAPPING& keyToProcess : keyMappings)
 	{
 		/* Skip inappropriate ones when necessary */
-		if (isIgnoredMapping(bExclude, *keyToProcess))
+		if (isIgnoredMapping(bExclude, bAllowMouseWheelEvents, keyToProcess))
 		{
 			continue;
 		}
 
-		// Check if there exists another keybinding, with the same subKeyCode, but with meta- or altMetaKeyCode
-		// and which has its respective meta key currently down. If any are found, then we want to completely skip
-		// any bindings without meta, as the keybind with meta (being more specific) should take precedence. This
-		// allows e.g. having camera movement in arrow keys and alignment in arrow keys + ctrl
-		const bool bIsKeyCombination = keyToProcess->metaKeyCode != KEY_IGNORE;
-		if (bMetaKeyDown && !bIsKeyCombination) {
-			bool bOverridingCombinationExists = false;
-			for (auto otherKey = keyMappings.begin(); otherKey != keyMappings.end(); ++otherKey)
-			{
-				/* Skip inappropriate ones when necessary */
-				if (isIgnoredMapping(bExclude, *keyToProcess))
-				{
-					continue;
-				}
-
-				/* Only match agaist keys with the same subKeyCode */
-				if (otherKey->subKeyCode != keyToProcess->subKeyCode)
-				{
-					continue;
-				}
-
-				bool bIsOtherKeyCombination = otherKey->metaKeyCode != KEY_IGNORE;
-				if (bIsOtherKeyCombination && keyPressed(otherKey->subKeyCode))
-				{
-					bool bHasAlt = otherKey->altMetaKeyCode != KEY_IGNORE;
-					if (keyDown(otherKey->metaKeyCode) || (bHasAlt && keyDown(otherKey->altMetaKeyCode)))
-					{
-						bOverridingCombinationExists = true;
-						break;
-					}
-				}
-			}
-
-			if (bOverridingCombinationExists) {
-				continue;
-			}
-		}
-
-		/* Process simple ones (single keys) */
-		if (!bIsKeyCombination)
+		/* Skip if the input is already consumed. Handles skips for context/priority and meta-conflicts */
+		const auto bIsAlreadyConsumed = std::find_if(consumedInputs.begin(), consumedInputs.end(), [keyToProcess](const KeyMappingInput& consumed) {
```

Comment:
Would a `std::set` work here?

Reply:
No idea why I was worried about insertion perf before; the find gets called much more often 👍 Changed this to use `std::unordered_map`